multibranch/src/main/java/org/jenkinsci/plugins/workflow/multibranch/WorkflowProjectFactoryImpl.java

```diff
@@ -0,0 +1,99 @@
+ *
+ * Copyright 2015 CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.jenkinsci.plugins.workflow.multibranch;
+
+import hudson.Extension;
+import hudson.model.Item;
+import java.io.IOException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import jenkins.branch.Branch;
+import jenkins.branch.BranchProjectFactory;
+import jenkins.branch.BranchProjectFactoryDescriptor;
+import jenkins.branch.MultiBranchProject;
+import org.jenkinsci.plugins.workflow.cps.CpsScmFlowDefinition;
+import org.jenkinsci.plugins.workflow.job.WorkflowJob;
+import org.jenkinsci.plugins.workflow.job.WorkflowRun;
+import org.kohsuke.stapler.DataBoundConstructor;
+
+public class WorkflowProjectFactoryImpl extends BranchProjectFactory<WorkflowJob,WorkflowRun> {
+    
+    private static final Logger LOGGER = Logger.getLogger(WorkflowProjectFactoryImpl.class.getName());
+
+    static final String SCRIPT = "jenkins.groovy";
+
+    @DataBoundConstructor public WorkflowProjectFactoryImpl() {}
+
+    @Override public WorkflowJob newInstance(Branch branch) {
+        WorkflowJob job = new WorkflowJob((WorkflowMultiBranchProject) getOwner(), branch.getName());
+        setBranch(job, branch);
+        return job;
+    }
+
+    @Override public Branch getBranch(WorkflowJob project) {
+        return project.getProperty(BranchJobProperty.class).getBranch();
+    }
+
+    @Override public WorkflowJob setBranch(WorkflowJob project, Branch branch) {
+        BranchJobProperty property = project.getProperty(BranchJobProperty.class);
+        try {
+            if (property == null) {
+                property = new BranchJobProperty();
+                setBranch(property, branch, project);
+                project.addProperty(property);
+            } else if (!property.getBranch().equals(branch)) {
+                setBranch(property, branch, project);
+                project.save();
+            }
```

Comment:
Would it really be ok for a `BranchJobProperty` to already have a branch name set on it? And would it be okay to change it?